### PR TITLE
Trait case sensitivity and aliasing collision fixes

### LINKED ISSUE #663: Changed trait method visibility is not taken into account
```php
<?php
trait HelloWorld {
    public function sayHello() {
        echo 'Hello World!';
    }
}

// Change visibility of sayHello
class MyClass1 {
    use HelloWorld { sayHello as protected; }
}

// Alias method with changed visibility
// sayHello visibility not changed
class MyClass2 {
    use HelloWorld { sayHello as private myPrivateHello; }
}
```

BetterReflection keeps the original visibility from the trait.

In the AST this information is communicated through `PhpParser\Node\Stmt\TraitUseAdaptation\Alias::$newModifier` (integer constants from `Class_` most likely). Looks like this property isn't read in the source code at all.

Original issue: https://github.com/phpstan/phpstan/issues/3465#issuecomment-643757521

### PR DESCRIPTION
**Pull request by kukulich**
Reported by @ondrejmirtes

Fixes https://github.com/Roave/BetterReflection/issues/663 and other bugs.

### CODE REVIEW

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -1019,16 +1030,44 @@ private function getTraitPrecedences(): array
...
+        $this->parseTraitUsages();
+
+        /** @return array<string, int> */
+        return $this->cachedTraitModifiers;
+    }
+
     private function parseTraitUsages(): void
     {
-        if ($this->cachedTraitAliases !== null && $this->cachedTraitPrecedences !== null) {
+        if ($this->cachedTraitAliases !== null && $this->cachedTraitModifiers !== null && $this->cachedTraitPrecedences !== null) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php:1062]: Having 3 different caches for the same thing is quite worrisome :D

File: src/Reflection/ReflectionClass.php [hunk 2]
```diff
@@ -236,11 +240,18 @@ private function createMethodsFromTrait(ReflectionMethod $method): array
...
         $traitPrecedences = $this->getTraitPrecedences();
+        $traitModifiers   = $this->getTraitModifiers();
 
         $methodAst = $method->getAst();
         assert($methodAst instanceof ClassMethod);
 
-        $methodHash   = $this->methodHash($method->getImplementingClass()->getName(), $method->getName());
+        $methodHash = $this->methodHash($method->getImplementingClass()->getName(), $method->getName());
+
+        if (isset($traitModifiers[$methodHash])) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Let's use `array_key_exists()` for these - `isset()` is just generally too risky for type-level regressions later on

File: src/Reflection/ReflectionClass.php [hunk 3]
```diff
@@ -1038,12 +1077,20 @@ private function parseTraitUsages(): void
             foreach ($adaptations as $adaptation) {
                 $usedTrait = $adaptation->trait;
                 if ($usedTrait === null) {
-                    $usedTrait = $traitNames[0];
+                    $usedTrait = end($traitNames);
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php:1080]: Did the index change in upstream `nikic/php-parser`?

File: test/unit/Reflection/ReflectionClassTest.php [hunk 4]
```diff
@@ -1882,4 +1882,122 @@ class Foo
...
+            {
+                protected function myMethod() : void{
+                
+                }
+            }
+            
+            class HelloWorld
+            {
+                use MyTrait {
+                    MyMethod as myRenamedMethod;
```
Comment (Ocramius) [inline test/unit/Reflection/ReflectionClassTest.php:1901]: Sneaky casing issue :O

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Reflection/ReflectionClass.php (hunk 3 above)
Comment (kukulich) [reply src/Reflection/ReflectionClass.php:1080]: No, we just need to read the latest.

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (kukulich) [reply src/Reflection/ReflectionClass.php:1062]: I will remove the new one and combine with one of the previous. It should be easy.

Comment (kukulich) [reply src/Reflection/ReflectionClass.php:1062]: Hmm, `getTraitAliases()` is `public` so it would be more difficult...

Comment (kukulich) [reply src/Reflection/ReflectionClass.php:1062]: I suggest to ignore it for now... :)

Comment (Ocramius) [reply src/Reflection/ReflectionClass.php:1062]: Let's leave it alone - wasn't suggesting action is needed, but rather warning about ever-raising complexity.

**Review by Ocramius — APPROVED** [review]
LGTM 👍

**Pull request merged**